clang/lib/CodeGen/CodeGenModule.cpp

```diff
@@ -4347,6 +4362,19 @@ void CodeGenModule::emitCPUDispatchDefinition(GlobalDecl GD) {
   }
 }
 
+/// Adds a declaration to the list of multi version functions if not present.
+void CodeGenModule::AddDeferredMultiVersionResolverToEmit(GlobalDecl GD) {
+  const auto *FD = cast<FunctionDecl>(GD.getDecl());
+  assert(FD && "Not a FunctionDecl?");
+
+  if (FD->isTargetVersionMultiVersion()) {
+    StringRef NamePrefix = getMangledName(GD).split('.').first;
```

Comment:
ISTM you want `getMangledNameImpl` here instead, with `OmitMultiVersionMangling = true`. Trying to split this on the first `.` seems suspicious, and might not play well with `asm("foo.bar.baz._Mlse")`

Reply:
Initially I was using `FD->getName()` as the key here but this wouldn't work for C++ mangled names. `getMangledName()` internally calls `getMangledNameImpl()` which returns the multi versioned name (this would be "foo.bar.baz._Mlse" in you example). The key needs to be the prefix "foo.bar.baz" so that we don't push every multi version of "foo.bar.baz" in the list `MultiVersionFuncs` to be processed by `emitMultiVersionFunctions()`. This would generate a resolver body with multiple copies of the basic blocks corresponding to each version. Hmm, I am not sure how to handle this example properly, perhaps with `rsplit()`.

Reply:
I tried experimenting with __asm__ but couldn't construct a sensible example in regards to multi versioning. Can you please show me?

Reply:
As far as I can tell `target_version` and `asm` cannot be combined according to clang's semantic checker, therefore a function declaration which has been renamed via `asm` cannot reach `AddDeferredMultiVersionResolverToEmit()`. In any case I will use `rsplit()` instead.

Reply:
`asm`+multiversioning isn't really the point of that argument... sorry, bad example. My point is that the `.` inserted for multi versioning mangling might not be the only one. If you insist on string splitting, it should at least be `rsplit`.

We shouldn't have to rely on string splitting though: `getMangledNameImpl(..., false)` returns the prefix that you need. It is used elsewhere in this multiversioning code to retrieve what the mangled name would be on the function if it did not have the attribute:

https://github.com/llvm/llvm-project/blob/7337db72ed334f8389601f160b762e50c4e61c25/clang/lib/CodeGen/CodeGenModule.cpp#L1857-L1861

Reply:
Right, I see. I've just thought of another issue with `getMangledNameImpl()`. Even though it would be more reliable to not split strings, we can't use its return value (std::string) as the key to a DenseSet. But let me double check it.

Reply:
Hm. Is there somewhere convenient to stick an `llvm::StringSaver`?

Reply:
Surprisingly it didn't seem necessary. I inserted the std::string to the DenseSet (which still expects a StringRef) and I am now running the Address sanitizer just to be sure :D

Reply:
Nopes, it didn't like it!